### PR TITLE
Prometheus: properly de-accumulate multi-heatmap responses

### PR DESCRIPTION
**Pull request by gtk-grafana**
mostly addresses https://github.com/grafana/support-escalations/issues/3373

**What this PR does / why we need it**:
We currently have a bug that is exhibited when a user has multiple prometheus queries that return values for the same label.
When we're calculating the heatmap values, all heatmaps (across queries) are merged into a single array and sorted by label. 
The heatmap value is calculated by subtracting subsequent values to get the delta in value between x-axis intervals.
When multiple queries are included that have the same labels (in the example below, the quartile `le`), the resulting calculation subtracts values in the same quartile across different queries resulting in the incorrect calculation.

Example bug:
image

Example after fix:
image

Using sample data:
```
test_histogram_bucket{additional_bucket="100",le="+Inf"} 7
test_histogram_bucket{additional_bucket="100",le="0.05"} 1
test_histogram_bucket{additional_bucket="100",le="0.1"} 2
test_histogram_bucket{additional_bucket="100",le="0.15"} 3
test_histogram_bucket{additional_bucket="100",le="0.2"} 4
test_histogram_bucket{additional_bucket="100",le="0.25"} 5
test_histogram_bucket{additional_bucket="100",le="0.3"} 6
test_histogram_bucket{additional_bucket="20",le="+Inf"} 26
test_histogram_bucket{additional_bucket="20",le="0.05"} 20
test_histogram_bucket{additional_bucket="20",le="0.1"} 21
test_histogram_bucket{additional_bucket="20",le="0.15"} 22
test_histogram_bucket{additional_bucket="20",le="0.2"} 23
test_histogram_bucket{additional_bucket="20",le="0.25"} 24
test_histogram_bucket{additional_bucket="20",le="0.3"} 25
test_histogram_bucket{additional_bucket="50",le="+Inf"} 16
test_histogram_bucket{additional_bucket="50",le="0.05"} 10
test_histogram_bucket{additional_bucket="50",le="0.1"} 11
test_histogram_bucket{additional_bucket="50",le="0.15"} 12
test_histogram_bucket{additional_bucket="50",le="0.2"} 13
test_histogram_bucket{additional_bucket="50",le="0.25"} 14
test_histogram_bucket{additional_bucket="50",le="0.3"} 15
```

**Which issue(s) this PR fixes**:
Fix for https://github.com/grafana/support-escalations/issues/3373

Fixes https://github.com/grafana/support-escalations/issues/3373

**Special notes for your reviewer**:
This should solve the reported bug in https://github.com/grafana/support-escalations/issues/3373!

### CODE REVIEW

File: public/app/plugins/datasource/prometheus/result_transformer.test.ts [hunk 1]
```diff
@@ -434,7 +434,7 @@ describe('Prometheus Result Transformer', () => {
                 name: 'Value',
                 type: FieldType.number,
                 values: [30, 10, 40],
-                labels: { le: '3', additionalProperty: '10' },
+                labels: { le: 'Inf+', additionalProperty: '10' },
```
Comment (leeoniya) [inline public/app/plugins/datasource/prometheus/result_transformer.test.ts:437]: oops? :)

Comment (gtk-grafana) [reply public/app/plugins/datasource/prometheus/result_transformer.test.ts:437]: Haha yes this is a typo, but the `parseSampleValue` should return the same result 🤔 but I guess we're already covering that in another test... I'll fix!

File: devenv/.gitignore [hunk 2]
```diff
@@ -0,0 +1 @@
+*.prom
```
Comment (leeoniya) [inline devenv/.gitignore:1]: let's exclude this file from the PR

File: public/app/plugins/datasource/prometheus/result_transformer.test.ts [hunk 3]
```diff
@@ -291,6 +291,245 @@ describe('Prometheus Result Transformer', () => {
...
+      } as unknown as DataQueryResponse;
+
+      const series = transformV2(response, options, {});
+      expect(series.data[0].fields.length).toEqual(4);
+      expect(series.data[0].fields[1].values.toArray()).toEqual([10, 10, 0]);
+      expect(series.data[0].fields[2].values.toArray()).toEqual([10, 0, 30]);
+      expect(series.data[0].fields[3].values.toArray()).toEqual([10, 0, 10]);
+    });
+
+    it('results with heatmap format with multiple values for each bucket should be correctly transformed', () => {
```
Comment (leeoniya) [inline public/app/plugins/datasource/prometheus/result_transformer.test.ts:392]: maybe?

```suggestion
    it('results with heatmap format and multiple histograms should be grouped and de-accumulated by remaining labels', () => {
```

**Comment by gtk-grafana** [comment]
Thanks to @leeoniya we were able to figure out a generic solution to this problem! Here's the same example data above in a single query, using the `Join by labels` transform! 🎉 

image
image

**Review by leeoniya — APPROVED** [review]
LGTM :shipit:

File: public/app/plugins/datasource/prometheus/result_transformer.test.ts [hunk 4]
```diff
@@ -291,6 +291,245 @@ describe('Prometheus Result Transformer', () => {
...
+      expect(series.data[0].fields[2].values.toArray()).toEqual([10, 0, 30]);
+      expect(series.data[0].fields[3].values.toArray()).toEqual([10, 0, 10]);
+
+      expect(series.data[1].fields[1].values.toArray()).toEqual([0, 10, 10]);
+      expect(series.data[1].fields[2].values.toArray()).toEqual([20, 0, 30]);
+      expect(series.data[1].fields[3].values.toArray()).toEqual([10, 0, 20]);
+
+      expect(series.data[2].fields[1].values.toArray()).toEqual([30, 30, 60]);
+      expect(series.data[2].fields[2].values.toArray()).toEqual([0, 10, 0]);
+      expect(series.data[2].fields[3].values.toArray()).toEqual([10, 0, 0]);
```
Comment (leeoniya) [inline public/app/plugins/datasource/prometheus/result_transformer.test.ts:530]: wonder if we should merge these into a single frame :thinking: cc @ryantxu 

would be a single frame with fields like:

```
Time
1
1
2
2
+Inf
+Inf
```

File: public/app/plugins/datasource/prometheus/result_transformer.ts [hunk 5]
```diff
@@ -102,9 +102,39 @@ export function transformV2(
...
+    // Get reference to dataFrames for heatmap
+    const heatmapResultsGroup = heatmapResultsGroupedByQuery[query];
+
+    // Create a new grouping by iterating through the data frames...
+    const heatmapResultsGroupedByValues = groupBy<DataFrame>(heatmapResultsGroup, (dataFrame) => {
+      // Each data frame has `Time` and `Value` properties, we want to get the values
+      const values = dataFrame.fields.find((field) => field.name === TIME_SERIES_VALUE_FIELD_NAME);
+      // Specific functionality for special "le" quantile heatmap value, we know if this value exists, that we do not want to calculate the heatmap density across data frames from the same quartile
+      if (values?.labels && HISTOGRAM_QUANTILE_LABEL_NAME in values.labels) {
+        const { le, ...notLE } = values?.labels;
```
Comment (gtk-grafana) [inline public/app/plugins/datasource/prometheus/result_transformer.ts:122]:
```suggestion
        const { le, ...notLE } = values.labels;
```

**Comment by bohandley** [comment]
I can see that it is grouping better! @gtk-grafana and @leeoniya this is looking better. Unfortunately, I am missing something. I cannot reproduce the visualization. It might be something to do with the time range. When I do a time range zoom out, the visualization looks similar to the expected results, but I am also getting doubled labels and negative results

1. Initial visualization
Screen Shot 2022-08-16 at 3 15 46 PM

2. Time range zoom out with repeated labels
Screen Shot 2022-08-16 at 3 17 00 PM

3. One more time range zoom out with negative values
Screen Shot 2022-08-16 at 3 17 45 PM

Here are the steps I am doing to test this. Attached at the bottom is a video of how I am creating the visualization.

1. Add data to devenv
     a. cd into devenv
     b. add the file `metrics.prom` to the devenv directory
     c. copy and paste the following data into `metrics.prom` with a newline at the end of the file
```
test_histogram_bucket{additional_bucket="100",le="+Inf"} 7
test_histogram_bucket{additional_bucket="100",le="0.05"} 1
test_histogram_bucket{additional_bucket="100",le="0.1"} 2
test_histogram_bucket{additional_bucket="100",le="0.15"} 3
test_histogram_bucket{additional_bucket="100",le="0.2"} 4
test_histogram_bucket{additional_bucket="100",le="0.25"} 5
test_histogram_bucket{additional_bucket="100",le="0.3"} 6
test_histogram_bucket{additional_bucket="20",le="+Inf"} 26
test_histogram_bucket{additional_bucket="20",le="0.05"} 20
test_histogram_bucket{additional_bucket="20",le="0.1"} 21
test_histogram_bucket{additional_bucket="20",le="0.15"} 22
test_histogram_bucket{additional_bucket="20",le="0.2"} 23
test_histogram_bucket{additional_bucket="20",le="0.25"} 24
test_histogram_bucket{additional_bucket="20",le="0.3"} 25
test_histogram_bucket{additional_bucket="50",le="+Inf"} 16
test_histogram_bucket{additional_bucket="50",le="0.05"} 10
test_histogram_bucket{additional_bucket="50",le="0.1"} 11
test_histogram_bucket{additional_bucket="50",le="0.15"} 12
test_histogram_bucket{additional_bucket="50",le="0.2"} 13
test_histogram_bucket{additional_bucket="50",le="0.25"} 14
test_histogram_bucket{additional_bucket="50",le="0.3"} 15

```
     d. Change the `node_exporter` configuration in `docker-compose.yaml`
```
  # node_exporter:
  #   image: prom/node-exporter
  #   ports:
  #     - "9100:9100"
  node_exporter:
    image: prom/node-exporter
    volumes:
      - ./metrics.prom:/prometheus/metrics/metrics.prom
    ports:
      - "9100:9100"
    command:
      - --collector.disable-defaults
      - --collector.textfile
      - --collector.textfile.directory=/prometheus/metrics
      - --log.level=debug
```
     e. In devenv directory, run `docker-compose up`
     f. Check http://localhost:9100/metrics that the data from `metrics.prom` was added.

2. Start up Grafana and create a new dashboard panel with the gdev-prometheus datasource
     a. query `test_histogram_bucket`
     b. Change `format: time series` to `format: heatmap` in the query options
     c. Select the barchart visualization
     d. Select the `last 5 minutes` time range
     e. Click the Transform tab and select the `Join by labels` transformation
          -value: `additional bucket`
          -join: `le`


https://user-images.githubusercontent.com/25674746/184964394-c9e9e2af-e0b8-4b45-8950-a619e3b38697.mov

**Comment by gtk-grafana** [comment]
@bohandley Sorry about that, apparently we neglected to mention that the graph `type` must be instant, which should fix the issue you're seeing here!

**Comment by bohandley** [comment]
You're right! I did miss that!

Start up Grafana and create a new dashboard panel with the gdev-prometheus datasource
a. query test_histogram_bucket
b. Change format: time series to format: heatmap in the query options
c. Select the barchart visualization
d. Select the last 5 minutes time range
e. Click the Transform tab and select the Join by labels transformation
-value: additional bucket
-join: le
***Fixed by adding
f. change custom labels in the options to {{le}}
g. Choose type: instant in the options


Screen Shot 2022-08-16 at 4 21 37 PM

**Comment by bohandley** [comment]
I looked at a few other implementations of format: heatmap, comparing the previous behavior of this with histogram queries and wasn't able to see any changes or side effects. LGTM!

**Review by bohandley — APPROVED** [review]

**Review by itsmylife — APPROVED** [review]
LGTM

**Pull request merged**